Add a --drop-variables flag to xray.open_dataset to exclude certain variables

markelg opened the pull request:
Related to issue #457. I implemented this flag following the instructions given by @shoyer in the issue thread. I have a decent amount of experience with python, but this is the first pull request I set up in GitHub, and I am a begginer with git (more used to svn). I was careful but please check that I did not mess up something ; )

shoyer commented on xray/conventions.py:
```diff
@@ -871,6 +874,9 @@ def decode_cf(obj, concat_characters=True, mask_and_scale=True,
     vars, attrs, coord_names = decode_cf_variables(
         vars, attrs, concat_characters, mask_and_scale, decode_times,
         decode_coords)
+    if drop_variables is not None:
```
could you move this into `decode_cf_variables` instead? probably could put this logic in the loop that constructs `new_vars`

shoyer commented:
Looks pretty good to me. However, this does need tests to verify that it works -- see here for some examples: https://github.com/xray/xray/blob/6ed84a04334338533c6773ce4b37d2179130df18/xray/test/test_conventions.py#L460-L493

Also, this needs a note in "What's New" in the docs.

shoyer commented on xray/backends/api.py:
```diff
@@ -114,6 +114,9 @@ def open_dataset(filename_or_obj, group=None, decode_cf=True,
         used when reading data from netCDF files with the netcdf4 and h5netcdf
         engines to avoid issues with concurrent access when using dask's
         multithreaded backend.
+    drop_variables: iterable, optional 
```
Maybe it would also be useful to support providing a single variable as a string? Someone is bound to try that. You can detect that with `isinstance(drop_variables, basestring)`.

markelg commented:
I updated it following your advice, now the logic is in the loop of decode_cf_variables, and it supports single strings. I added the test and the documentation too. I successfully run nosetests.

markelg commented:
I am not able to reproduce the error found by Travis on a virtualenv with the same python version and packages, so I don't know why that specific run failed : (

jhamman commented:
@markelg - The failing test you have is failing for the same reason as #529.  I don't think it is related to your PR.  @shoyer - Thoughts on what's going on here?

shoyer commented:
I'll take a look -- something probably changed in one of our upstream dependencies.

On Mon, Aug 17, 2015 at 9:06 AM, Joe Hamman <notifications@github.com>
wrote:

> @markelg - The failing test you have is failing for the same reason as #529.  I don't think it is related to your PR.  @shoyer - Thoughts on what's going on here?
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/xray/xray/pull/532#issuecomment-131874475

shoyer commented on xray/conventions.py:
```diff
@@ -807,6 +807,10 @@ def stackable(dim):
 
     new_vars = OrderedDict()
     for k, v in iteritems(variables):
+        if isinstance(drop_variables, basestring):
```
let's do this outside the loop

shoyer replied:
Let's also convert `drop_variables` to a set. That will keep things performant, even if the dataset has a very large number of variables.

shoyer commented on xray/backends/api.py:
```diff
@@ -114,6 +114,10 @@ def open_dataset(filename_or_obj, group=None, decode_cf=True,
         multithreaded backend.
+    drop_variables: string or iterable, optional 
+        A variable or list of variables to exclude from being parsed from the
+        dataset.This may be useful to drop variables with problems or
```
missing a space after the end of the sentence

shoyer commented:
@markelg the test failure was unrelated (see #535 for details). If you merge or rebase on the latest master your tests should be passing.

shoyer commented on xray/test/test_conventions.py:
```diff
@@ -491,6 +491,21 @@ def test_decode_cf_with_multiple_missing_values(self):
+            'y': ('t', [5, 10, np.nan])
+        })
+        actual = conventions.decode_cf(original, drop_variables=("x",))
+        self.assertDatasetIdentical(expected, actual)
```
let's also test providing a single argument, `drop_variables='x'`

shoyer commented on xray/conventions.py:
```diff
@@ -804,9 +804,13 @@ def stackable(dim):
 
     coord_names = set()
+    if isinstance(drop_variables, basestring):
+        drop_variables = set([drop_variables,])
```
actually -- we should always convert drop_variables to a set. I was thinking something like:

```python
if isinstance(drop_variables, basestring):
    drop_variables = [drop_variables]
elif drop_variables is None:
   drop_variables = []
drop_variables = set(drop_variables)
```

shoyer replied:
this would even let us skip `drop_variables is None` below.

shoyer commented:
OK, merging. Thanks @markelg !
